### PR TITLE
Feature: Azure Key Vault

### PR DESCRIPTION
**Pull request by vg-svitla**

### CODE REVIEW

File: Azure/azure-key-vault/ingest/parser.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  - name: json_event
+    external:
+      name: json.parse-json
+  - name: set_ecs_fields
+
+stages:
+  set_ecs_fields:
+    actions:
+      - set:
+          event.kind: "event"
```
Comment (squioc) [inline Azure/azure-key-vault/ingest/parser.yml:12]: We start withdrawing `event.kind: "event"` as most events in the platform are `event` (But we continue to keep this field for alerts and other kind).
So, In this case, I suggest removing this field:
```suggestion
```

File: Azure/azure-key-vault/ingest/parser.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          event.category: "database"
```
Comment (squioc) [inline Azure/azure-key-vault/ingest/parser.yml:13]: Must be an array:
```suggestion
          event.category: ["database"]
```

File: Azure/azure-key-vault/ingest/parser.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+          event.dataset: "keyvault"
+          event.outcome: "{{json_event.message.resultSignature}}"
```
Comment (squioc) [inline Azure/azure-key-vault/ingest/parser.yml:15]: The value of `event.outcome` is restricted: https://www.elastic.co/guide/en/ecs/8.6/ecs-allowed-values-event-outcome.html
I suggest setting this field according to the resultSignature: (`failure` for `Forbidden`, `success` for `OK`).

File: Azure/azure-key-vault/ingest/parser.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+          event.action: "{{json_event.message.operationName}}"
+          event.type: ["access", "change", "info"]
+          cloud.provider: "azure"
```
Comment (squioc) [inline Azure/azure-key-vault/ingest/parser.yml:18]: I think we can add [`cloud.service.name`](https://www.elastic.co/guide/en/ecs/8.6/ecs-cloud.html#field-cloud-service-name):
```suggestion
          cloud.provider: "azure"
          cloud.service.name: "Azure Key Vault"
```

File: Azure/azure-key-vault/ingest/parser.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+          url.original: "{{json_event.message.properties.requestUri}}"
+          tls.version: "{{json_event.message.properties.tlsVersion}}"
+          azure.key_vault.result.type: "{{json_event.message.resultType}}"
+          azure.key_vault.result.description: "{{json_event.message.resultDescription}}"
+          azure.key_vault.result.signature: "{{json_event.message.resultSignature}}"
+          azure.key_vault.correlation_id: "{{json_event.message.correlationId}}"
+          azure.key_vault.resource_id: "{{json_event.message.resourceId}}"
+
+      - set:
+          event.type: ["denied", "change", "info"]
```
Comment (squioc) [inline Azure/azure-key-vault/ingest/parser.yml:32]: `denied` doesn't match the `database` category: https://www.elastic.co/guide/en/ecs/8.6/ecs-allowed-values-event-category.html#ecs-event-category-database
I suggest using only one type:
```suggestion
          event.type: ["access"]
```

File: Azure/azure-key-vault/ingest/parser.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+stages:
+  set_ecs_fields:
+    actions:
+      - set:
+          event.kind: "event"
+          event.category: "database"
+          event.dataset: "keyvault"
+          event.outcome: "{{json_event.message.resultSignature}}"
+          event.action: "{{json_event.message.operationName}}"
+          event.type: ["access", "change", "info"]
```
Comment (squioc) [inline Azure/azure-key-vault/ingest/parser.yml:17]: I suggest using only one type:
```suggestion
          event.type: ["access"]
```

**Review by squioc — CHANGES_REQUESTED** [review]
Thanks for the job.
Some changes to apply.

File: Azure/azure-key-vault/ingest/parser.yml (hunk 4 above)
Comment (vg-svitla) [reply Azure/azure-key-vault/ingest/parser.yml:17]: Should we also add
```
          observer.vendor: "Microsoft"
          observer.product: "Azure Key Vault"
```
?

File: Azure/azure-key-vault/_meta/fields.yml [hunk 7]
```diff
@@ -0,0 +1,24 @@
+azure.key_vault.correlation_id:
```
Comment (Adamowoc) [inline Azure/azure-key-vault/_meta/fields.yml:1]: Could you create custom fields for the fields in identity.claim and the fields in properties ?

**Review by Adamowoc — CHANGES_REQUESTED** [review]

File: Azure/azure-key-vault/_meta/fields.yml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+azure.key_vault.properties.enabledForDiskEncryption:
+  description: Indicates whether the key vault is enabled for disk encryption
+  name: azure.key_vault.properties.enabledForDiskEncryption
+  type: boolean
+
+azure.key_vault.properties.enabledForTemplateDeployment:
+  description: Indicates whether the key vault is enabled for template deployment
+  name: azure.key_vault.properties.enabledForTemplateDeployment
+  type: boolean
```
Comment (squioc) [inline Azure/azure-key-vault/_meta/fields.yml:40]: I doubt about the usefulness of these fields.
I suggest removing them
```suggestion
```
What do you think @Adamowoc?

File: Azure/azure-key-vault/_meta/fields.yml [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  description: The SKU name of the key vault
+  name: azure.key_vault.properties.sku.name
+  type: keyword
+
+azure.key_vault.properties.softDeleteRetentionInDays:
+  description: The number of days that items are retained in the soft delete area
+    before being permanently deleted
+  name: azure.key_vault.properties.softDeleteRetentionInDays
+  type: number
+
```
Comment (squioc) [inline Azure/azure-key-vault/_meta/fields.yml:62]: Same as above
```suggestion
```

File: Azure/azure-key-vault/ingest/parser.yml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+          azure.key_vault.identity.claim.appid: "{{json_event.message.identity.claim.appid}}"
+          azure.key_vault.properties.tenantid: "{{json_event.message.properties.properties.tenantId}}"
+          azure.key_vault.properties.softDeleteRetentionInDays: "{{json_event.message.properties.properties.softDeleteRetentionInDays}}"
+          azure.key_vault.properties.enabledForDeployment: "{{json_event.message.properties.properties.enabledForDeployment}}"
+          azure.key_vault.properties.enabledForDiskEncryption: "{{json_event.message.properties.properties.enabledForDiskEncryption}}"
+          azure.key_vault.properties.enabledForTemplateDeployment: "{{json_event.message.properties.properties.enabledForTemplateDeployment}}"
+          azure.key_vault.properties.enableSoftDelete: "{{json_event.message.properties.properties.enableSoftDelete}}"
+          azure.key_vault.properties.enableRbacAuthorization: "{{json_event.message.properties.properties.enableRbacAuthorization}}"
+          azure.key_vault.properties.enablePurgeProtection: "{{json_event.message.properties.properties.enablePurgeProtection}}"
```
Comment (squioc) [inline Azure/azure-key-vault/ingest/parser.yml:38]: To match the suggestions:
```suggestion
```

**Review by squioc — APPROVED** [review]
LGTM.
I just wonder about the usefulness of some fields.

**Review by squioc — DISMISSED** [review]
LGTM

**Review by squioc — DISMISSED** [review]

File: Azure/azure-key-vault/tests/test_event_1.json [hunk 11]
```diff
@@ -0,0 +1,78 @@
+{
```
Comment (Adamowoc) [inline Azure/azure-key-vault/tests/test_event_1.json:1]: Could you give more precise names to the first 4 tests ?

File: Azure/azure-key-vault/_meta/fields.yml (hunk 7 above)
Comment (Adamowoc) [reply Azure/azure-key-vault/_meta/fields.yml:1]: @vg-svitla
Please add `identity.claim.scp` and` identity.claim.amr `
There is also more fields in properties such as the `id`, fields in `keyProperties`, `certificateProperties`, `secretProperties`, `isAccessPolicyMatch`, `tenantId`, etc. Please check `VaultPatch`, `CertificateImport (OK)` and `CertificateCreate` events.

**Review by squioc — APPROVED** [review]

File: Azure/azure-key-vault/ingest/parser.yml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+          azure.key_vault.correlation_id: "{{json_event.message.correlationId}}"
+          azure.key_vault.resource_id: "{{json_event.message.resourceId}}"
+          azure.key_vault.identity.claim.appid: "{{json_event.message.identity.claim.appid}}"
+          azure.key_vault.properties.tenantid: "{{json_event.message.properties.properties.tenantId}}"
+          azure.key_vault.properties.sku.family: "{{json_event.message.properties.properties.sku.Family}}"
+          azure.key_vault.properties.sku.name: "{{json_event.message.properties.properties.sku.Name}}"
+          azure.key_vault.properties.sku.capacity: "{{json_event.message.properties.properties.sku.Capacity}}"
+
+      - set:
+          event.outcome: "failure"
```
Comment (Adamowoc) [inline Azure/azure-key-vault/ingest/parser.yml:37]: We should also add "denied" to event.type

**Review by Adamowoc — CHANGES_REQUESTED** [review]

File: Azure/azure-key-vault/_meta/fields.yml (hunk 7 above)
Comment (vg-svitla) [reply Azure/azure-key-vault/_meta/fields.yml:1]: Ok seems like I did not understand the description of task in correct way. Will fix it

File: Azure/azure-key-vault/ingest/parser.yml (hunk 12 above)
Comment (squioc) [reply Azure/azure-key-vault/ingest/parser.yml:63]: I agree

**Review by squioc — CHANGES_REQUESTED** [review]
Some duplicated fields to remove

File: Azure/azure-key-vault/ingest/parser.yml (hunk 12 above)
Comment (vg-svitla) [reply Azure/azure-key-vault/ingest/parser.yml:62]: I did answer in DM, but will duplicate here
Seems like in ECS we do not have `denied` for `database` and for `outcome`
https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-outcome.html
https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html#ecs-event-category-database

**Review by squioc — DISMISSED** [review]
LGTM

**Review by Adamowoc — CHANGES_REQUESTED** [review]
In order to reduce the size of the logs in DB we will be removing some fields

**Review by Adamowoc — APPROVED** [review]
Thanks for the great job

**Pull request merged**